### PR TITLE
Assignment7+8 linq - Hunter Thompson & Hunter Allen

### PR DESCRIPTION
**Pull request by StrangeRanger**
@hallen

### CODE REVIEW

**Comment by FenrisStargazer** [comment]
Will review.

File: Assignment/Assignment.Tests/SampleDataTest.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        Assert.IsTrue(data.Contains(expected));
+        Assert.IsFalse(data.Contains(notExpected));
+    }
+
+    // TODO: Check this test...
+    [TestMethod]
+    public void GetPeople_AreEqualIsTrue()
+    {
+        Assert.AreEqual(sampleData.CsvRows.Count(), sampleData.People.Count());
```
Comment (FenrisStargazer) [inline Assignment/Assignment.Tests/SampleDataTest.cs:105]: AreEqual < T >

**Review by FenrisStargazer — COMMENTED** [review]
**Instructions**
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌?✔
Using LINQ, skip the first row in the People.csv. ✔
_Technically yes, I think? But perhaps not the most efficient way of doing so._
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ❌?✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

Given the implementation of Node in Assignment5
Implement IEnumerable<T> to return all the items in the "circle" of items. ❌?✔
_So I am unsure if this is to say implement the **interface** IEnumerable<T> (which is what I am thinking) in which case this would be a ❌. However, if it simply means implement some method that utilizes the **type** IEnumerable<T> to do this, then this is a ✔. Clarification is needed._
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

**Extra Credit**
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

**Fundamentals**
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
_LangVersion missing_
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
_One instance where it is missing, noted with a comment._
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,55 @@
...
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
-
-        // 2.
-        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
-
-        // 3.
+        public IEnumerable<string> CsvRows { get; } = 
+            from string line in File.ReadAllLines("People.csv")
+            where !line.Contains("Id,FirstName,LastName,Email,StreetAddress,City,State,Zip")
```
[the remark this answers is not in the document]
Comment (StrangeRanger) [reply Assignment/Assignment/SampleData.cs:13]: Yes, but that's not what the instructions ask us to do.
"Using LINQ, skip the first row in the People.csv"

File: Assignment/Assignment.Tests/SampleDataTest.cs (hunk 1 above)
Comment (StrangeRanger) [reply Assignment/Assignment.Tests/SampleDataTest.cs:105]: I wasn't sure what the type would be for this one. I was hoping whoever was reviewing the code would potentially know.

File: Assignment/Assignment/SampleData.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (FenrisStargazer) [reply Assignment/Assignment/SampleData.cs:13]: I thought the Skip(int) method was still part of LINQ?

File: Assignment/Assignment.Tests/SampleDataTest.cs (hunk 1 above)
Comment (FenrisStargazer) [reply Assignment/Assignment.Tests/SampleDataTest.cs:105]: I believe it is an int given that it's a count function. But if you wanted to be sure you could look up the docs for the method, it usually lists the return type.

File: Assignment/Assignment/SampleData.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (StrangeRanger) [reply Assignment/Assignment/SampleData.cs:13]: Huh, looks like it is. I'll go ahead and use it then. Thanks for making the comment here.

**Review by mmwoodfo — COMMENTED** [review]
Good job! Only comment is to stick to either query syntax or the more c# like linq syntax

**Pull request closed without merging**